pkg/urbit/vere/epoc.c

```diff
@@ -0,0 +1,903 @@
+            "           try again using the latest version of the urbit\r\n"
+            "           binary. If you have questions or concerns, please\r\n"
+            "           file an issue against\r\n"
+            "           https://github.com/urbit/urbit and assign it to\r\n"
+            "           mcevoypeter.\r\n");
+    goto rename_lock_mdb;
+  }
+
+  MDB_txn* txn_u;
+  { // Read metadata out of LMDB instance.
+    try_lmdb(mdb_txn_begin(poc_u->env_u, NULL, MDB_RDONLY, &txn_u),
+             goto rename_lock_mdb,
+             "failed to create read-only transaction");
+
+    MDB_dbi dbi_u;
+    try_lmdb(mdb_dbi_open(txn_u, "META", 0, &dbi_u),
+             goto abort_txn,
+             "failed to open META database");
+
+#define lookup_metadata(key_string, destination)                               \
+    do {                                                                       \
+      MDB_val key_u = {                                                        \
+        .mv_data = (void*)key_string,                                          \
+        .mv_size = strlen(key_string) + 1,                                     \
+      };                                                                       \
+      MDB_val val_u;                                                           \
+      try_lmdb(mdb_get(txn_u, dbi_u, &key_u, &val_u),                          \
+               goto abort_txn,                                                 \
+               "failed to lookup metadata value for key '%s'",                 \
+               key_string);                                                    \
+      destination = u3i_bytes(val_u.mv_size, val_u.mv_data);                   \
+    } while ( 0 )
+
+    u3_atom who;
+    lookup_metadata("who", who);
+    u3r_chubs(0, 2, met_u->who_d, who);
+    u3z(who);
+    lookup_metadata("fake", met_u->fak_o );
+    lookup_metadata("life", met_u->lif_w);
+    lookup_metadata("version", met_u->ver_w);
+#undef lookup_metadata
+
+    mdb_txn_abort(txn_u);
+  }
+
+  { // Write metadata to binary files.
+    c3_path_push(poc_u->pax_u, ver_nam_c);
+    // Convert to network byte order to ensure portability across platforms of
+    // varying endianness.
+    c3_w ver_w = htonl(met_u->ver_w);
+    if ( !c3_bile_write_new(poc_u->pax_u, &ver_w, sizeof(ver_w)) ) {
+      goto rename_lock_mdb;
+    }
+    c3_path_pop(poc_u->pax_u);
+
+    c3_path_push(poc_u->pax_u, lif_nam_c);
+    c3_w lif_w = htonl(met_u->lif_w);
+    if ( !c3_bile_write_new(poc_u->pax_u, &lif_w, sizeof(lif_w)) ) {
+      goto rename_lock_mdb;
+    }
+    c3_path_pop(poc_u->pax_u);
+  }
+
+  goto succeed;
+
+abort_txn:
+  mdb_txn_abort(txn_u);
+rename_lock_mdb:
+  c3_assert(_move_file(poc_u->pax_u, src_u, "lock.mdb"));
+rename_data_mdb:
+  c3_assert(_move_file(poc_u->pax_u, src_u, "data.mdb"));
+free_epoc:
+  u3_epoc_close(poc_u);
+  c3_free(poc_u);
+fail:
+  return NULL;
```

Comment:
I'm concerned about this approach to failed migrations. Moving files first -- and moving them back if anything is missing/incomplete or if anything goes wrong -- is more fragile than collecting and validating all data before moving the files. I/O can fail, and if it is at all possible, we want to leave the pier undisturbed when we can't migrate.

Reply:
This has also been corrected.